### PR TITLE
Change move encoding

### PR DESCRIPTION
**Pull request by ceebo**
This patch changes the move encoding. I believe it is nicer and also faster. The main idea is to have the top bit represent promotions _and_ castlings. We differentiate between promotions and castlings by looking at the to/from squares. I think the trick I found for doing this is nice.

Doing it this way has the advantage that for castling/promotion there are now 3 free bits to encode more data. For promotions we can use the simplest possible encoding of piece type and for castlings we can use the extra bit to signify gating on the rook square. I wrote detailed comments about this in types.h.

The patch changes bench but again I claim this is for trivial reasons. If you take master and change search.cpp line 595 to:

```
    posKey = pos.key() ^ Key(12345); // was Key(excludedMove)
```

then master and the equivalent version of this branch give the same bench number. Having done this I did a make profile-build and did speed test between this branch and master with ./stockfish bench 16 1 20. On average over 3 runs the gain in speed was close to 2%.

### CODE REVIEW

File: src/position.cpp [hunk 1]
```diff
@@ -888,18 +888,11 @@ void Position::do_move(Move m, StateInfo& newSt, bool givesCheck) {
...
 
-  // Update incremental scores
-  st->psq += PSQT::psq[pc][to] - PSQT::psq[pc][from];
-
-  // Set capture piece
-  st->capturedPiece = captured;
-
-  // Add gating piece
-  if (is_gating(m))
+  else if (gating_type(m) != NO_GATE) // Safe because m is not a pawn move
```
Comment (ceebo) [inline src/position.cpp]: I suppose I cheated by adding this optimisation that is not related to the move encoding. Still I doubt this accounts for all of the speed gain I saw.

**Comment by ianfab** [comment]
Interesting. It took a while for me to understand how the magic around bit 3 works, but now the move encoding makes sense to me. It will probably be not until next weekend that I have the time and focus for a full review.

**Comment by ceebo** [comment]
Yes it is a bit tricky I admit. Especially that once the "(m & CASTLING) ?" test is passed I rely on the fact that the new expression "m ^ (m >> 6)" will also have the CASTLING bit set.

I thought of a way that is the same number of operations but possibly a bit simpler to understand. It would be like this:

```
NORMAL = 0
ENPASSANT = 1 << 14
PROMOTION = (1<<16)-1
CASTLING = PROMOTION ^ 8
```
and then
```
type_of(m) = m & (1 << 15) ? CASTLING | (m ^ (m >> 6)) : m & ENPASSANT; 
```

So PROMOTION has all 16 bits set while CASTLING is the same but with bit 3 cleared. If the parities of the ranks are different then we add bit 3 into CASTLING to turn it into PROMOTION.

I am relaxed about hearing your comments, don't feel like there is any rush. Cheers.

EDIT: Doing it this way would shift some pain towards the make_move function. The best I can think of is 

```
return std::min(T, 1<<15) + (pt_encode << 12) + (from << 6) + to;
```
for line 482 of types.h because I never want T to cause the setting of both bits 14 and 15.

EDIT2: Yuck, if T is CASTLING2 then I _do_ want bits 14 and 15 set. Sigh... this way is starting to not look nicer at all...

EDIT3: Last edit ;) With the 4 types as above we could have `CASTLING2 = 3 << 14` and then:

```
template<MoveType T>
inline Move make(Square from, Square to, PieceType pt = NO_GATE) {
  MoveType Flags = (T == PROMOTION || T == CASTLING) ? (1 << 15) : T;
  PieceType pt_encode = pt - (T == PROMOTION ? NO_PIECE_TYPE : NO_GATE);
  return Move(Flags + (pt_encode << 12) + (from << 6) + to);
}
```

**Comment by ianfab** [comment]
@ceebo Thanks, but I think your first version is fine. However, although you already thoroughly explain the move encoding in the one long comment, it might help to add a few short comments/reminders at the crucial points in the code. I will add review comments on that.

File: src/types.h [hunk 2]
```diff
@@ -131,11 +138,10 @@ enum Move : int {
...
 enum MoveType {
   NORMAL,
-  PROMOTION  = 1 << 14,
-  ENPASSANT  = 2 << 14,
-  CASTLING   = 3 << 14,
-  PROMOTION2 = ENPASSANT,
-  CASTLING2  = PROMOTION
+  ENPASSANT = 1 << 14,
+  CASTLING  = 1 << 15,
+  PROMOTION = CASTLING | 8,
```
Comment (ianfab) [inline src/types.h]: I think a short comment that only the CASTLING part is used for the actual encoding would be helpful here.

In my opinion it would also make sense to add an alias like `CASTLING_OR_PROMOTION` in order to avoid confusion, since you can more clearly differentiate between the "fake" values and the ones that are XORed exactly as they are without stripping bits.

File: src/types.h [hunk 3]
```diff
@@ -438,41 +445,41 @@ inline int from_to(Move m) {
  return m & 0xFFF;
 }
 
-inline bool is_gating(Move m) {
-  return (m & (3 << 12)) && (rank_of(from_sq(m)) == RANK_1 || rank_of(from_sq(m)) == RANK_8) && ((m & (3 << 14)) == NORMAL || (m & (3 << 14)) == CASTLING || rank_of(from_sq(m)) == rank_of(to_sq(m)));
+inline MoveType type_of(Move m) {
+  return MoveType(m & CASTLING ? (m ^ (m >> 6)) & PROMOTION : m & ENPASSANT);
```
Comment (ianfab) [inline src/types.h]: This is a place where a flag like `CASTLING_OR_PROMOTION` might help to avoid confusion.

File: src/types.h [hunk 4]
```diff
@@ -438,41 +445,41 @@ inline int from_to(Move m) {
...
 
-inline MoveType type_of(Move m) {
-  if ((m & (3 << 14)) == PROMOTION2 && (m & (3 << 12)))
-      return PROMOTION;
-  if (gating_on_castling_rook(m))
-      return CASTLING;
-  return MoveType(m & (3 << 14));
+inline PieceType promotion_type(Move m) {
+  assert(type_of(m) == PROMOTION);
+  return PieceType((m >> 12) - (CASTLING >> 12));
```
Comment (ianfab) [inline src/types.h]: `(m ^ CASTLING) >> 12` (or `(m ^ CASTLING_OR_PROMOTION) >> 12`) might be simpler.

File: src/types.h (hunk 3 above)
Comment (ceebo) [reply src/types.h]: I agree. The problem with this is that CASTLING_OR_PROMOTION would be visible throughout the entire source but would only have meaning within types.h. I could do as follows:

```
#define CASTLING_OR_PROMOTION (1 << 12)
#define FLAGS (3 << 12)
#define GATE_ON_TO (1 << 13)
```
and then undef them at the end of the file. Do you think this is a good approach?

EDIT: I think I lean towards improving readability of this function by adding comments instead of adding extra flag definitions. You?

**Comment by ceebo** [comment]
Thanks for the comments.

I pushed another commit with some minor changes:

1. Moved the piece encoding to the highest bits. This saves a couple of operations when decoding and the only cost is one extra line (with run-time cost 0) in make_move.

2. Added a couple of comments in the MoveType enum definitions.

3. Changed NO_GATE to NO_GATE_TYPE. It seems to fit better with NO_PIECE_TYPE.

Regarding CASTLING_OR_PROMOTION I answered next to one of your comments.

**Comment by ianfab** [comment]
Thanks, the changes look good. Adding a short comment to type_of would indeed help, and since it now is the only place left where the CASTLING flag could potentially be confusing, a comment definitely makes more sense than an additional flag.

**Comment by ceebo** [comment]
Great. I added a comment explaining _what_ type_of returns. I'm not sure I can make it completely obvious _how_ it does this though. Seems like it will always be something that takes a little time to understand.

After making the change Key(excludedMove) -> Key(12345) in search.cpp I checked that I was still getting the same bench as previously. This seems to imply I made no obvious blunders with the most recent commits.

**Comment by ianfab** [comment]
Ok, I think the PR should be ready for merge now, or do you have any pending changes?

**Comment by ceebo** [comment]
I don't have anything pending. Happy to see this merged.
